src/unitxt/metrics.py

```diff
@@ -6012,76 +6014,265 @@ def compute(
         return {self.main_score: [len(prediction)], "score_name": self.main_score}
 
 
-class GraniteGuardianWMLMetric(InstanceMetric):
+class RiskType(str, Enum):
+    """Risk type for the Granite Guardian models"""
+
+    RAG = "rag_risk"
+    USER_MESSAGE = "user_risk"
+    ASSISTANT_MESSAGE = "assistant_risk"
+    AGENTIC = "agentic_risk"
+    CUSTOM_RISK = "custom_risk"
+
+
+class GraniteGuardianMetric(InstanceMetric):
     """Return metric for different kinds of "risk" from the Granite-3.0 Guardian model."""
 
-    main_score = "granite_guardian"
     reduction_map: Dict[str, List[str]] = None
     prediction_type = float
+    main_score = None
+    reduction_map = {}
+    wml_model_name: str = "ibm/granite-guardian-3-8b"
+
+    wml_params = {
+        "decoding_method": "greedy",
+        "max_new_tokens": 20,
+        "temperature": 0,
+        "return_options": {
+            "top_n_tokens": 5,
+        },
+    }
+
+    hf_model_name: str = "ibm-granite/granite-guardian-3.1-8b"
 
-    model_name: str = "ibm/granite-guardian-3-8b"
-    hf_model_name: str = "ibm-granite/granite-guardian-3.0-8b"
     safe_token = "No"
     unsafe_token = "Yes"
 
-    inference_engine: WMLInferenceEngineGeneration = None
+    inference_engine: LogProbInferenceEngine = None
     generation_params: Dict = None
     risk_name: str = None
+    risk_type: RiskType = None
+    risk_definition: Optional[str] = None
+
+    user_message_field: str = "user"
+    assistant_message_field: str = "assistant"
+    context_field: str = "context"
+    tools_field: str = "tools"
+
+    available_risks: Dict[RiskType, List[str]] = {
+        RiskType.USER_MESSAGE: [
+            "harm",
+            "social_bias",
+            "jailbreak",
+            "violence",
+            "profanity",
+            "unethical_behavior",
+        ],
+        RiskType.ASSISTANT_MESSAGE: [
+            "harm",
+            "social_bias",
+            "violence",
+            "profanity",
+            "unethical_behavior",
+        ],
+        RiskType.RAG: ["context_relevance", "groundedness", "answer_relevance"],
+        RiskType.AGENTIC: ["function_call"],
+    }
 
-    _requirements_list: List[str] = ["ibm_watsonx_ai", "torch", "transformers"]
+    _requirements_list: List[str] = ["torch", "transformers"]
 
-    def prepare(self):
-        self.reduction_map = {"mean": [self.main_score]}
+    def verify_guardian_config(self, task_data):
```

Comment:
The following lines are very complicated. Can we instead use Inheritance? So we have `BaseGuerdianGraniteMetric`
Then `RagGuerdianGraniteMetric` where minimal things are changed? Not sure if possible, but I think it will might make this code section as well as it usage more simple

Reply:
I think it makes sense, I will implement it.

Reply:
I am having second thoughts on this. I agree on the fact that the code is complex, but adding one inheritence per risk type makes it harder for users to use in case they don't want to use the catalog and they create the classes directly.

Reply:
@elronbandel done